Forge.TreeWalker/src/SubroutineAction.cs

```diff
@@ -0,0 +1,113 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="SubroutineAction.cs" company="Microsoft">
+//     Copyright (c) Microsoft Corporation.  All rights reserved.
+// </copyright>
+// <summary>
+//     The SubroutineAction class.
+// </summary>
+//-----------------------------------------------------------------------
+
+namespace Forge.TreeWalker
+{
+    using System;
+    using System.Threading.Tasks;
+
+    using Forge.Attributes;
+
+    /// <summary>
+    /// The SubroutineAction is a native Forge action that walks a TreeWalkerSession.
+    /// This action performs the following steps:
+    ///   1. Rehydrates the Subroutine's SessionId if previously persisted.
+    ///   2. Gets an initialized TreeWalkerSession for this Subroutine from the App, passing in the SubroutineInput from the schema and Subroutine SessionId.
+    ///   3. Explicitly calls UpdateKeyPrefix on the ForgeState with the updated Subroutine SessionId.
+    ///   4. Walks the tree starting at the CurrentTreeNode if persisted, otherwise RootTreeNodeKey if given, otherwise "Root" as default.
+    ///   5. Returns the Subroutine's last ActionResponse if available, otherwise returns the tree walker session Status.
+    /// </summary>
+    [ForgeAction(InputType: typeof(SubroutineInput))]
+    public class SubroutineAction : BaseAction
+    {
+        /// <summary>
+        /// The tree walker parameters of the parent tree walker session.
+        /// Used to call InitializeSubroutineTree to get the initialized TreeWalkerSession for this Subroutine.
+        /// </summary>
+        private TreeWalkerParameters parameters { get; set; }
+
+        /// <summary>
+        /// Instantiates a SubroutineAction with the required parameters.
+        /// </summary>
+        /// <param name="parameters">The tree walker parameters of the parent tree walker session.</param>
+        public SubroutineAction(TreeWalkerParameters parameters)
+        {
+            this.parameters = parameters ?? throw new ArgumentNullException(nameof(parameters));
+        }
+
+        /// <summary>
+        /// The RunAction method is called when Forge encounters an ActionName while walking the tree.
+        /// </summary>
+        /// <param name="actionContext">The action context holding relevant information for this Action.</param>
+        public override async Task<ActionResponse> RunAction(ActionContext actionContext)
+        {
+            SubroutineInput input = (SubroutineInput)actionContext.ActionInput;
+
+            // Rehydrate the subroutine's SessionId if previously persisted.
+            SubroutineIntermediates intermediates = await actionContext.GetIntermediates<SubroutineIntermediates>();
+            if (intermediates == null)
+            {
+                intermediates = new SubroutineIntermediates()
+                {
+                    SessionId = Guid.NewGuid()
+                };
```

Comment:
>; [](start = 17, length = 1)

nit: move this ";" to the end of line above

Reply:
I need the semi-colon here to end the line creating the SubroutineIntermediates object. I do not need semi-colon after initializing each parameter in this way (you use commas if there are multiple parameters).

---
In reply to: [356863479](https://github.com/microsoft/Forge/pull/6#discussion_r356863479) [](ancestors = 356863479)